src/hotspot/share/utilities/events.hpp

```diff
@@ -414,16 +428,46 @@ inline void EventLogBase<ExtendedStringLogMessage>::print(outputStream* out, Ext
   out->cr();
 }
 
+typedef void (*EventLogFunction)(Thread* thread, const char* format, ...);
+
+class EventMarkBase : public StackObj {
+  EventLogFunction _log_function;
+  StringLogMessage _buffer;
+
+ protected:
+  void log_start(const char* format, va_list argp) ATTRIBUTE_PRINTF(2, 0);
+  void log_end();
+
+ public:
+  EventMarkBase(EventLogFunction log_function);
+};
+
 // Place markers for the beginning and end up of a set of events.
-// These end up in the default log.
-class EventMark : public StackObj {
+template <EventLogFunction log_function>
+class EventMarkWithLogFunction : public EventMarkBase {
   StringLogMessage _buffer;
 
  public:
   // log a begin event, format as printf
-  EventMark(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
+  EventMarkWithLogFunction(const char* format, ...) ATTRIBUTE_PRINTF(2, 3) :
+      EventMarkBase(log_function) {
+    if (LogEvents) {
+      va_list ap;
+      va_start(ap, format);
+      log_start(format, ap);
+      va_end(ap);
+    }
+  }
   // log an end event
-  ~EventMark();
+  ~EventMarkWithLogFunction() {
+    if (LogEvents) {
+      log_end();
+    }
+  }
 };
 
+// These end up in the default log.
+typedef EventMarkWithLogFunction<Events::log> EventMark;
+typedef EventMarkWithLogFunction<Events::log_vm_operation> EventMarkVMOperation;
```

Comment:
Can you add a comment just above EventMarkVMOperation line 470 like,
// These go to the vm_operation log
For some reason, I read the comment as applying to both instantiations.

Reply:
Thanks for reviewing.